test/gateway/eip20_gateway/progress_unstake_with_proof.js

```diff
@@ -0,0 +1,511 @@
+      true,
+      `Reward amount ${receivedReward.toString(10)} must be less than ${maxReward.toString(10)}`,
+    );
+    assert.strictEqual(
+      eventData._rewardAmount.eq(receivedReward),
+      true,
+      `Reward received in the event ${eventData._rewardAmount.toString(10)} should be same as actual received reward ${receivedReward}`,
+    );
+
+  });
+
+  it('should fail when message hash is zero', async function () {
+
+    let messageHash = ZeroBytes;
+    await gateway.setInboxStatus(unstakeMessage.messageHash, MessageStatusEnum.Declared);
+    let blockHeight = await setStateRoot(gateway);
+    let storageProof = StubData.co_gateway.redeem.proof_data.storageProof[0].serializedProof;
+
+    await Utils.expectRevert(
+      gateway.progressUnstakeWithProof(
+        messageHash,
+        storageProof,
+        blockHeight,
+        MessageStatusEnum.Declared
+      ),
+      'Message hash must not be zero.'
+    );
+
+  });
+
+  it('should fail when unstake message is undeclared', async function () {
+
+    await gateway.setInboxStatus(unstakeMessage.messageHash, MessageStatusEnum.Undeclared);
+    let blockHeight = await setStateRoot(gateway);
+    let storageProof = StubData.co_gateway.redeem.proof_data.storageProof[0].serializedProof;
+
+    await Utils.expectRevert(
+      gateway.progressUnstakeWithProof(
+        unstakeMessage.messageHash,
+        storageProof,
+        blockHeight,
+        MessageStatusEnum.Declared
+      ),
+      'Message on target must be Declared.',
+    );
+  });
+
+  it('should fail when unstake message is already progressed', async function () {
+
+    await gateway.setInboxStatus(unstakeMessage.messageHash, MessageStatusEnum.Declared);
+    let blockHeight = await setStateRoot(gateway);
+    let storageProof = StubData.co_gateway.redeem.proof_data.storageProof[0].serializedProof;
+
+    await gateway.progressUnstakeWithProof(
+      unstakeMessage.messageHash,
+      storageProof,
+      blockHeight,
+      MessageStatusEnum.Declared
+    );
+
+    await Utils.expectRevert(
+      gateway.progressUnstakeWithProof(
+        unstakeMessage.messageHash,
+        storageProof,
+        blockHeight,
+        MessageStatusEnum.Declared
+      ),
+      'Message on target must be Declared.',
+    );
+  });
+
+  it('should fail for revoked redeem(unstake) message', async function () {
+
+    await gateway.setInboxStatus(unstakeMessage.messageHash, MessageStatusEnum.Revoked);
+    let blockHeight = await setStateRoot(gateway);
+    let storageProof = StubData.co_gateway.redeem.proof_data.storageProof[0].serializedProof;
+
+    await Utils.expectRevert(
+      gateway.progressUnstakeWithProof(
+        unstakeMessage.messageHash,
+        storageProof,
+        blockHeight,
+        MessageStatusEnum.Declared
+      ),
+      'Message on target must be Declared.',
+    );
+
+  });
+
+  it('should fail when the reward amount is greater than the unstake amount', async function () {
```

Comment:
General question: this means only a revert is possible?

Reply:
Depends, if another chain has progressed than revert is also not possible. To solve this, a check is added during stake and redeem which ensure stake/redeem amount should always be greater than maximum reward. 
Refer https://github.com/OpenSTFoundation/mosaic-contracts/issues/568

Reply:
So this test case cannot happen in practice? 😕

Reply:
Yes, As an integration it's not possible. As a unit it is.

Reply:
ok